compiler/frontend/src/org/jetbrains/kotlin/resolve/calls/inference/ConstraintSystemImpl.kt

```diff
@@ -101,33 +101,36 @@ internal class ConstraintSystemImpl(
         override fun hasTypeInferenceIncorporationError() = errors.any { it is TypeInferenceError } || !satisfyInitialConstraints()
 
         override fun hasTypeParameterWithUnsatisfiedOnlyInputTypesError() =
-            localTypeParameterBounds.values.any { it.typeVariable.hasOnlyInputTypesAnnotation() && it.value == null }
+            allTypeParameterBounds.entries.any {
+                !it.key.isExternal && it.value.typeVariable.hasOnlyInputTypesAnnotation() && it.value.value == null
+            }
 
         override val constraintErrors: List<ConstraintError>
             get() = errors
     }
 
     private fun getParameterToInferredValueMap(
-        typeParameterBounds: Map<TypeVariable, TypeBoundsImpl>,
+        typeParameterBounds: SmartIdentityTable<TypeVariable, TypeBoundsImpl>,
         getDefaultType: (TypeVariable) -> KotlinType,
         substituteOriginal: Boolean
-    ): Map<TypeConstructor, TypeProjection> {
-        val substitutionContext = HashMap<TypeConstructor, TypeProjection>()
-        for ((variable, typeBounds) in typeParameterBounds) {
+    ): SmartIdentityTable<TypeConstructor, TypeProjection> {
+        val substitutionContext = SmartIdentityTable<TypeConstructor, TypeProjection>()
+        for (variable in typeParameterBounds.keys) {
+            val typeBounds = typeParameterBounds[variable]!!
             val value = typeBounds.value
             val typeConstructor =
                 if (substituteOriginal) variable.originalTypeParameter.typeConstructor
                 else variable.type.constructor
             val type =
                 if (value != null && !TypeUtils.contains(value, DONT_CARE)) value
                 else getDefaultType(variable)
-            substitutionContext.put(typeConstructor, TypeProjectionImpl(type))
+            substitutionContext[typeConstructor] = TypeProjectionImpl(type)
         }
         return substitutionContext
     }
 
     override val typeVariables: Set<TypeVariable>
```

Comment:
Note that behavior was sneakily changed here in a similar way: a new collection is returned instead of a view. This is probably not a big deal, but we could've changed the `ConstraintSystem` interface instead so that `typeVariables` returned a `Sequence` to keep the old behavior